syft/file/secrets_search_by_line_strategy.go

```diff
@@ -0,0 +1,123 @@
+package file
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"regexp"
+
+	"github.com/anchore/syft/syft/source"
+)
+
+func catalogLocationByLine(resolver source.FileResolver, location source.Location, patterns map[string]*regexp.Regexp) ([]SearchResult, error) {
+	readCloser, err := resolver.FileContentsByLocation(location)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch reader for location=%q : %w", location, err)
+	}
+	defer readCloser.Close()
+
+	var scanner = bufio.NewReader(readCloser)
+	var position int64
+	var allSecrets []SearchResult
+	var lineNo int64
+	var readErr error
+	for !errors.Is(readErr, io.EOF) {
+		lineNo++
+		var line []byte
+		// TODO: we're at risk of large memory usage for very long lines
+		line, readErr = scanner.ReadBytes('\n')
+		if readErr != nil && readErr != io.EOF {
+			return nil, readErr
+		}
+
+		lineSecrets, err := searchForSecretsWithinLine(resolver, location, patterns, line, lineNo, position)
+		if err != nil {
+			return nil, err
+		}
+		position += int64(len(line))
+		allSecrets = append(allSecrets, lineSecrets...)
+	}
+
+	return allSecrets, nil
+}
+
+func searchForSecretsWithinLine(resolver source.FileResolver, location source.Location, patterns map[string]*regexp.Regexp, line []byte, lineNo int64, position int64) ([]SearchResult, error) {
+	var secrets []SearchResult
+	for name, pattern := range patterns {
+		matches := pattern.FindAllIndex(line, -1)
+		for i, match := range matches {
+			if i%2 == 1 {
+				// FindAllIndex returns pairs of numbers for each match, we are only interested in the starting (first)
+				// position in each pair.
+				continue
+			}
+
+			lineOffset := int64(match[0])
+			secret, err := extractSecretFromPosition(resolver, location, name, pattern, lineNo, lineOffset, position+lineOffset)
+			if err != nil {
+				return nil, err
+			}
+			if secret != nil {
+				secrets = append(secrets, *secret)
+			}
+		}
+	}
+
+	return secrets, nil
+}
+
+func extractSecretFromPosition(resolver source.FileResolver, location source.Location, name string, pattern *regexp.Regexp, lineNo, lineOffset, seekPosition int64) (*SearchResult, error) {
```

Comment:
This function looks large and complex. What are your thoughts about decomposing it?

Reply:
I'll split out the "seeking" logic from the "finder" logic to help out with complexity. (I don't recommend we break apart the "finder" logic into multiple parts.)